test/jdk/java/net/httpclient/FileChannelPublisherTest.java

```diff
@@ -0,0 +1,693 @@
+            assertArrayEquals(expectedRequestBodyBytes, actualRequestBodyBytes);
+
+        });
+
+    }
+
+    /**
+     * <em>Big enough</em> file length to observe the effects of publisher state corruption while uploading.
+     * <p>
+     * Certain tests follow below steps:
+     * </p>
+     * <ol>
+     * <li>Issue the request</li>
+     * <li>Wait for the server's signal that the request (not the body!) is received</li>
+     * <li>Corrupt the publisher's state; modify the file, close the file channel, etc.</li>
+     * <li>Signal the server to proceed with reading</li>
+     * </ol>
+     * <p>
+     * With small files, even before we permit the server to read (step 4), file gets already uploaded.
+     * This voids the effect of state corruption (step 3).
+     * To circumvent this, use this <em>big enough</em> file size.
+     * </p>
+     *
+     * @see #testChannelCloseDuringPublisherRead(ServerRequestPair, Path)
+     * @see #testFileModificationDuringPublisherRead(ServerRequestPair, Path)
+     */
+    private static final int BIG_FILE_LENGTH = 8 * 1024 * 1024;  // 8 MiB
+
+    @ParameterizedTest
+    @MethodSource("serverRequestPairs")
+    void testChannelCloseDuringPublisherRead(
+            ServerRequestPair pair,
+            @TempDir(cleanup = CleanupMode.ON_SUCCESS) Path tempDir)
+            throws Exception {
+        establishInitialConnection(pair);
+        ServerRequestPair.SERVER_REQUEST_RECEIVED_SIGNAL = new CountDownLatch(1);
+        ServerRequestPair.SERVER_READ_PERMISSION = new CountDownLatch(1);
+        try {
+
+            int fileLength = BIG_FILE_LENGTH;
+            AtomicReference<CompletableFuture<HttpResponse<Void>>> responseFutureRef = new AtomicReference<>();
+            withFileChannel(tempDir.resolve("data.txt"), fileLength, ((_, fileChannel) -> {
+
+                // Issue the request
+                LOGGER.log("Issuing the request");
+                HttpRequest request = pair
+                        .requestBuilder
+                        .POST(ofFileChannel(fileChannel, 0, fileLength))
+                        .build();
+                responseFutureRef.set(CLIENT.sendAsync(request, discarding()));
+
+                // Wait for server to receive the request
+                LOGGER.log("Waiting for the request to be received");
+                ServerRequestPair.SERVER_REQUEST_RECEIVED_SIGNAL.await();
+
+            }));
+
+            LOGGER.log("File channel is closed");
+
+            // Let the server proceed
+            LOGGER.log("Permitting the server to proceed");
+            ServerRequestPair.SERVER_READ_PERMISSION.countDown();
+
+            // Verifying the client failure
+            LOGGER.log("Verifying the client failure");
+            Exception requestFailure = assertThrows(ExecutionException.class, () -> responseFutureRef.get().get());
+            assertInstanceOf(UncheckedIOException.class, requestFailure.getCause());
```

Comment:
I think the application should be receiving a `IOException` (both for send() and sendAsync()) instead of a `UncheckedIOException`.

Reply:
The `IOE` thrown is wrapped by an `UncheckedIOE` in `FileChannelIterator::next`, which overrides `Iterator::next` and that does not allow a `throws` in the `next()` footprint. Would you mind elaborating on your remark, please?

Reply:
> add a test which uses send()

@jaikiran, in faa5d24d831, I've adapted all exception verification tests to take both `send()` and `sendAsync()` into account.

>  both send() and sendAsync() should propagate the same exception

Right. I will discuss this with @dfuch and @Michael-Mc-Mahon.